iree/hal/local/sync_semaphore.h

```diff
@@ -0,0 +1,75 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IREE_HAL_LOCAL_SYNC_SEMAPHORE_H_
+#define IREE_HAL_LOCAL_SYNC_SEMAPHORE_H_
+
+#include "iree/base/api.h"
+#include "iree/base/synchronization.h"
+#include "iree/hal/api.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+//===----------------------------------------------------------------------===//
+// iree_hal_sync_semaphore_state_t
+//===----------------------------------------------------------------------===//
+
+// State shared between all sync semaphores.
+// Owned by the device and guaranteed to remain valid for the lifetime of any
+// semaphore created from it.
+typedef struct {
+  // In-process notification signaled when any semaphore value changes.
+  iree_notification_t notification;
+} iree_hal_sync_semaphore_state_t;
+
+// Initializes state used to perform semaphore synchronization.
+void iree_hal_sync_semaphore_state_initialize(
+    iree_hal_sync_semaphore_state_t* out_shared_state);
+
+// Deinitializes state used to perform semaphore synchronization; no semaphores
+// must be live with references.
+void iree_hal_sync_semaphore_state_deinitialize(
+    iree_hal_sync_semaphore_state_t* shared_state);
+
+//===----------------------------------------------------------------------===//
+// iree_hal_sync_semaphore_t
+//===----------------------------------------------------------------------===//
+
+// Creates a semaphore that allows for ordering of operations on the local host.
```

Comment:
Can you elaborate a bit more on this (what a "sync semaphore" is)? This is backed by a 400LOC `.c`, which is only a bit less than `task_semaphore.c`. Seems worth a comparison.

Reply:
Added a note -- semaphores are mechanisms to propagate errors, query for completion, etc and are useful whether synchronously executing work or not, and having them in both cases lets us have a consistent API to the application (and make use of them in the compiled output) without having to special case things. Even on bare-metal systems using the synchronous driver they may still have multiple cores and want to query the status of work from one core to another - semaphores are a way to do that in a thread-safe fashion (even if your threads are not traditional threads but independent execution units). On super bare-metal systems with single cores/single threaded usage the iree_slim_mutex_t/iree_notification_t types can be no-oped, in which case this becomes just a way to carry around an integer payload and a status field.